### PR TITLE
[rllib] Pull out the GPU-parallel optimizer from policy gradients to common

### PR DESCRIPTION
**Pull request by ericl**
Pulls out the GPU-parallel SGD optimizer from policy gradients into rllib common. This might be useful for parallelizing DQN next, and is of general use for parallel optimization.

cc @pcmoritz @atumanov

### CODE REVIEW

File: python/ray/rllib/parallel.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+  parallel. The gradients are then averaged and applied to the shared weights.
+
+  The data loaded is pinned in device memory until the next call to
+  `load_data`, so you can make multiple passes (possibly in randomized order)
+  over the same data once loaded.
+
+  This is similar to tf.train.SyncReplicasOptimizer, but works within a single
+  TensorFlow graph, i.e. implements in-graph replicated training:
+
+    https://www.tensorflow.org/versions/r0.12/how_tos/distributed/
```
Comment (pcmoritz) [inline python/ray/rllib/parallel.py]: let's point to https://www.tensorflow.org/api_docs/python/tf/train/SyncReplicasOptimizer here, the above link is from 0.12 which is pretty old by now

Comment (ericl) [inline python/ray/rllib/parallel.py]: Done

**Comment by pcmoritz** [comment]
Tried it out on EC2 and it works on 2 gpus with 2x speedup.
